File: PYTHON/21_MergeTwoSortedLists.py
# Name: Aditya Chache
# Date: 05/10/2022


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:

        newnode = ListNode()
        # create a new linked list
        current_node = newnode
        # node to update

        while list1 and list2:
            if list1.val < list2.val:
                current_node.next = list1
                list1 = list1.next
            else:
                current_node.next = list2
                list2 = list2.next

            current_node = current_node.next


        # after loop has ended if one of the list is still having elements then add
        # those elements to the end of our list
        if list1:  
            current_node.next = list1
        elif list2:
            current_node.next = list2


        # return the new list
        return newnode.next 

File: C++/21_MergeTwoSortedLists.cpp
// Name: Md Farhan Sajid
// Date: 03/10/2022

class Solution {
  public:
    ListNode * mergeTwoLists(ListNode * l1, ListNode * l2) {
      if (l1 == NULL) {
        return l2;
      }

      if (l2 == NULL) {
        return l1;
      }

      ListNode * head = NULL;

      if (l1 -> val < l2 -> val) {
        head = l1;
        l1 = l1 -> next;
      } else {
        head = l2;
        l2 = l2 -> next;
      }

      ListNode * p;
      p = head;

      while (l1 && l2) {
        if (l1 -> val < l2 -> val) {
          p -> next = l1;
          l1 = l1 -> next;
        } else {
          p -> next = l2;
          l2 = l2 -> next;
        }

        p = p -> next;
      }

      if (l1 != NULL) {
        p -> next = l1;
      } else {
        p -> next = l2;
      }

      return head;
    }
};
